### ISSUE TITLE
Otter Pattern Library 

### ISSUE THREAD

**Issue by JohnPixle** [Codeinwp/otter-blocks#1203]
Hey team, as discussed on the Monday meeting, I worked on the first batch of Patterns made with Otter Blocks, in order to test them in the context of the Core Pattern library (both sidebar and core popup).

### Notes

- The aim for the Otter Patterns is to be cross-theme compatible, meaning that they should inherit the styles of the themes they are being used on, as much as possible. 
- Currently the images used are from the public domain, however I want to confirm if this restriction / requirement really applies. I am asking because the patterns that are included in the Neve theme are NOT from the public domain or CC0.  Let's clarify on the next meeting.
- Additionally, during this phase we can identify challenges, and see what we can improve in our blocks in order for the styling inheritance to take place as gracefully as possible. 
- I will provide feedback in a separate issue regarding the challenges met, and how we can improve the whole thing by making simple changes on some of the blocks.
- Overall it turned out to be an exceptionally challenging task, with endless fine-tuning required since every theme handles many things differently (especially core themes), from image alignments, block spacing and more. Doing an extensive research in the industry and competition, our patterns shall offer a good balance of simplicity and style inheritance, requiring just basic fine-tuning from the user.
- Next, the plan is to explore the integration of a custom Library Popup that will facilitate the use of Pro Patterns and page templates. It would be also good to create Pro Maintenance and coming soon templates, so we can cross-promote them in the WP Maintenance mode plugin.
- I have so far tested with the following themes:
     - Neve
     - Raft
     - Twenty Twenty
     - Twenty Twenty One
     - Twenty Twenty Two

### Patterns

The patterns are initially created in this demosite and page: https://demosites.io/otter/patterns/
For simplicity and testing efficiency I added all patterns in a single page.

To test the themes I eventually created a Taste instance: https://crazyotter.s1-tastewp.com/pattern-test/ 
If anyone wants to test, feel free to do so. Simply activate any of the available themes and refresh the page above.

```
 Admin area URL: https://crazyotter.s1-tastewp.com/wp-admin 
 Username: pixle 
 Password: 3uUjYXSkrDA 
```

✳️ Below is a link to the video walkthrough of how the pattern styles adapt when switching themes. Please take some time to watch it and let me know your input. We can evaluate some things and discuss more on the Monday meeting.

https://www.loom.com/share/188a9ba45739457d9b3bfe77b166aebf

Here is a screenshot of all patterns together, in order for you to have a quick overview
👇🏻👇🏻👇🏻

![Otter Patterns batch 1](https://user-images.githubusercontent.com/4193672/190656660-11f417bb-51bb-4ff3-a7e4-6b3b07c121e5.png)

**Comment by HardeepAsrani** [comment]
@JohnPixle Where did you take the images from? Can you link the credits?

**Comment by JohnPixle** [comment]
@HardeepAsrani Yes, I will do that. Here's one [quick example](https://www.rawpixel.com/image/1054569/post-impressionism-artwork-henri-rousseau) though. You can see the licensing (Public domain / CC0) in the sidebar.

**Comment by HardeepAsrani** [comment]
@JohnPixle Awesome, a beautiful selection too.

**Comment by selul** [comment]
maybe some review oriented patterns too? or they are too niched?

Awesome work, John !

**Comment by ineagu** [comment]
Hey! Looking nice.

@JohnPixle I am curious on why do you think they should inherit the theme styles? I think Elementor teached us that what people care more is that the templates looks the same to any theme.

**Comment by JohnPixle** [comment]
Thanks for the input everyone.

@selul 
> maybe some review oriented patterns too? or they are too niched?

This is absolutely in the plan, but I would wait until the updated version of the review block rolls out in the next update. We have made significant improvements to it. Also, I would consider just one review-based pattern to be free and offer more review pattern variants in the PRO version. However I can add a review-based pattern in the collection this week, since at this stage we also need to see how patterns with complex blocks get previewed.

@ineagu 

> I am curious on why do you think they should inherit the theme styles? I think Elementor taught us that what people care more is that the templates looks the same to any theme.

I'd strongly debate this to be honest :) Here are my thoughts:

### Elementor
In the context of Elementor, the damage was done in the early days due to the lack of efficient Global Controls in the platform, among many other things. 

After Elementor Theme Styles were introduced, the consensus in the entire El community has shifted towards proper inheritance of styles, and avoiding to apply inline styles on the elements. 

Every serious Template Kit now in Elementor (even the core ones) avoid applying inline styles and rely more on inheriting styles from a parent source. 

Is this the right approach? Absolutely. Both in terms of performance, consistency and flexibility.
In its simplest form, if you have 20 Typographic elements on a page, does it make sense to specify the font family to every single one of them in CSS? Or define the family just once in the Theme Styles, and let every element inherit it? Definitely the second.

Also, since we are talking about patterns (which are more granular layout components), let's consider a  user who already has a Neve website with Inter Font defined in the customiser for body and headings. Now let's say the user wants to extend / scale-up some pages and layouts by inserting Otter patterns. If typography would not be inherited, users would need to go to every single Typographic element on these patterns, and reset the style in order to inherit the Inter font and be consistent with the rest of the site. This falls in the scope of efficiency in the design management process.

### WordPress

In core WordPress things are even more clear on what the best practices are, especially with Gutenberg and FSE (and lately with the focus on Style Variations). I would not go into details here, but the industry standard now is to have styles adapt to the theme's stylistic context. 

A good example is [Extendify](https://extendify.com/library/), which is a popular pattern library with style inheritance as a main selling point.

Additionally, so far, in Neve Starter Sites and Templates cloud we have followed an approach that resonates with the above concept. Avoid inline styles as much as we can and have elements inheriting styles from the customiser.


Sorry for the long message, but thought it would be useful to share our approach and mentality for other team members as well, in case they go through this message. This explains the logic of why we have been advocating for Style inheritance and reliance on default styles wherever possible, especially in the direction WP themes are moving.

Let me know your thoughts, and looking forward to hear anyone's input on this 👍🏻

**Comment by ineagu** [comment]
@JohnPixle It totally makes sense to aim for that, but it really depends on how usable we can make those work on most of the themes & you made a good point on patterns (wireframes) vs templates.

For me the patterns should still work & look well on 99% of the themes, be it popular or custom, old & new. Since the patterns are fairly simple this seems doable.

However I would question the previews here, won't make sense to go with black/white colors & very generic fonts, if they won't look like that in any of the cases, due to inheritance?

Going further with this logic, I guess the point I am trying to make is that no matter what theme the user is using the expectation is that once a pattern/template important it would look like the picture, otherwise is rather confusing.

So can we have dynamic previews? if not won't very wireframe like previews work better in setting the right expectations?

**Comment by JohnPixle** [comment]
@ineagu Of course. The patterns are not "tied" to a specific style, **the pattern preview is dynamic** so what users will see in the preview always depends on the theme they are using. Only the images remain the same. Font families and button colours are inherited by the theme, and this is to justify the flexibility and usability we both outlined in my previous comments. I think we are on the same page here actually.

There is no "default" style of the patterns. The screenshot I shared in my original comment is just for demonstrating the layout / structure. The style adapts always. 

Here is a screenshot of the same pattern in different themes.

![contextual](https://user-images.githubusercontent.com/4193672/190984118-be898db5-d0f5-408e-b82f-b181d9fc62a9.png)

### Wireframes

I like wireframes, and if you notice the patterns, they are already "wireframe-friendly". There are **no opinionated colors** (other than site bg / body text / and accent), and structurally they are very basic. One thing that is "opinionated" though, is the images. So if we replace them with gray / neutral boxes then we can achieve a much more wireframe-ish look. 

I am not sure if this the direction we want to take though, and if we want this to be more part of Otter or the TCP product.

In the next days I am going to integrate the patterns in the Core pattern library so we can see the actual preview and test everything in a real-life workflow context. I will share links for a tasteWP instance so we can all test-drive and evaluate the overall UIX. I would encourage everyone to take a look at the original video where you can see how the pattern styles adaptt to theme switching.


Please do not hesitate to ping me for any clarifications.
Many thanks!

**Comment by mghenciu** [comment]
Super nice work on the Patterns @JohnPixle ! 👨‍🎨 
I really like the images you used, they go well with the Default 2022 theme, and the whole look of the Theme.

___
I support the idea of **having better color defaults for the Icon Block & the Border**.
This will also help us if we other blocks that rely on this, for example the Accordion block.

___
Regarding the wireframes idea, we discussed about this even in the previous Q, and we are thinking about adding Wireframes in Templates Cloud as a separate tab, and maybe we could also integrate some of them in Otter, with the option to get more of them by installing TPC.

**Comment by JohnPixle** [comment]
@mghenciu thanks, I would then summarise the changes we need to do into the following:

- [ ] Across our **border controls,** instead of having an inline custom value, we should let them inherit the body text color (whatever that is, depending on the theme). IF it takes too much time to apply this across all our blocks, I would prioritise it to apply to the Section block for now and specifically to the columns.

- [ ] Advanced icons (icon color AND border) should also by default inherit the body text color of the theme. 

The above two will allow our patterns to work well both in the context of light + dark themes.

@HardeepAsrani do you find any technical issues with this? If not I can create a separate issue for you guys to engage with these changes whenever you can.

**Comment by HardeepAsrani** [comment]
@JohnPixle A separate issue with more details will be helpful.

**Comment by JohnPixle** [comment]
# Integration and preview of Otter patterns in the Core library

As discussed in the last meeting, next step would be to integrate the Otter patterns into the core pattern library (both in the sidebar and the popup). Things went well and I am briefly sharing the results here.

### Notes

- I used the [Block Pattern Builder](https://wordpress.org/plugins/block-pattern-builder/) plugin to convert the Otter sections into actual patterns in order to test how they are rendered in the preview. The plugin is a bit outdated (2 years), so there were some issues (naturally). For example, there appears to be a limit in the number of custom patterns that can be displayed in the sidebar (and in popup). We will see how to bypass this, but this is not related to our patterns. For now I tested with a limited number of patterns, but it was enough to perform the tests.

- Apart from some minor issues in rendering, things are working fine. The previews are dynamic and the patterns adapt their styles based on the theme that is being used. 

- Any found issues we can / must address in Otter, will be reported in a separate issue in Github. I am currently working on that and will update you accordingly.

- I created a new Taste WP instance for anyone who wants to switch themes and test the patterns [in this page](https://overdofield.s3-tastewp.com/patterns-test/).

```
 Admin area URL: https://overdofield.s3-tastewp.com/wp-admin 
 Username: pixle 
 Password: YocxYuWKOHs 
```

### Sidebar Inserter and Popup

Screenshots below show the preview in both the pattern inserter sidebar (core) and the pattern popup (core)

![Frame 8](https://user-images.githubusercontent.com/4193672/191589852-05de08e1-956a-4598-ab6c-814cb9c79d7c.png)
![Frame 13](https://user-images.githubusercontent.com/4193672/191589865-23674f0a-e4db-4529-b935-1b41b21619b4.png)
![Frame 14](https://user-images.githubusercontent.com/4193672/191589875-57a1a5fb-f704-4346-a620-b9f4493abcbf.png)
![Frame 12](https://user-images.githubusercontent.com/4193672/191589899-f6384ecd-4892-492c-ae78-b8a9cb8d3d7a.png)


### Block Inserter

Also, the previews work fine in the block inserter as well. For example if the user is typing for icon block, he will also see the related otter patterns, with the correct styles.

![Frame 15](https://user-images.githubusercontent.com/4193672/191592692-2b8a37b1-0a75-4f76-9e13-2fc8118ea457.png)

cc @HardeepAsrani @mghenciu

**Comment by mghenciu** [comment]
I think the Patterns look good in the Inserter, nice job!

What do you think @JohnPixle , if we add a 10-20px spacer between the 2 progress bars we have?  so they don't look stacked on each other in preview (_I know that when imported, those look good and have some space between_):
![Screenshot 2022-09-22 at 12 25 35](https://user-images.githubusercontent.com/52494172/191711098-2dc57b2e-2a18-4ef8-9405-92953d75a7f8.png)

**Comment by JohnPixle** [comment]
@mghenciu Yeah, good point actually. The reason I didn't so far is because the spacing between these is changing depending on the theme's defined block gap. But I will better add a 10px divider anyway to play it safe.

**Comment by JohnPixle** [comment]
@HardeepAsrani I have completed the final tweaks on patterns, added a couple of new ones today (+1 testimonial, + author block, +2 content blocks).

I think we are good for a proper integration in the Core Sidebar and see/test how they work.

What is the best way for me to prepare them for you to register? You can grab the block codes from this page, where they are all together: https://demosites.io/otter/patterns but if you have other suggestions, let me know.

Now the issue is with naming the patterns (which is important for filtering and inserting). What would be efficient for you? I can simply add the names of the patterns here, based on the order in which they appear on the page. Sounds good?

If so, here is the list:

- [ ] Hero area with button
- [ ] Content with progress bars
- [ ] Text with image columns
- [ ] Large quote
- [ ] Image and text over dark background
- [ ] Contact details and form
- [ ] Call to action
- [ ] Testimonial columns
- [ ] Testimonial with inline image
- [ ] Centered testimonial with star icons
- [ ] Cover boxes with title and button
- [ ] Columns with image features
- [ ] Team members
- [ ] Columns with icon features
- [ ] Columns with flip boxes
- [ ] Icons and text
- [ ] Gallery
- [ ] Border icon features
- [ ] Border Pricing table
- [ ] Countdown
- [ ] Service boxes on dark background
- [ ] Content with features
- [ ] Author box

Let me know if you need any tweaks or further help with this. Please let's do an internal test first before we officially release them, as we will need to take a look at how they work in the core sidebar and the core popup inserter, and evaluate the whole performance.

Thanks!

**Comment by HardeepAsrani** [comment]
@JohnPixle @mghenciu I've made a PR with all these patterns: https://github.com/Codeinwp/otter-blocks/pull/1255 - let me know what you think. 😄

**Comment by JohnPixle** [comment]
@HardeepAsrani Thanks for the update, looking great 🚀 

I only see a minor issue with the alignment of some centered images, however it is not consistent with every theme... I am not sure if it is worth looking deeper into it, or simply leave it as is... Must be something related to each theme's CSS.

Let me know what you think, what I can do about this is to simply left-align the content in those patterns where alignment breaks. 

![Frame 16](https://user-images.githubusercontent.com/4193672/194811541-12b33cd9-8f44-484c-b53a-4b55570f8384.png)


Also, one thing I noticed when testing patterns in Neve, is that the body text is not previewed in the right font family. But I guess this something I must report to Neve repo.

![Screenshot 2022-10-10 at 09 55 15](https://user-images.githubusercontent.com/4193672/194812110-cf4cccdb-45e8-4f8d-9459-4a4e9b830459.png)


Thanks for the update again, looking forward to see he patterns released 🚀

**Comment by HardeepAsrani** [comment]
@JohnPixle Thanks, we can investigate the image issue but I'll merge it meanwhile as I don't see anything being wrong with patterns here.

**Comment by mghenciu** [comment]
Did some quick testing and overall things look very good 🚀 
With your permission John, I suggest to adjust a bit the paragraph text size on mobile for the First Header section (from 32 to 24px), **as right now both of them have 32 px on mobile:**
<img width="558" alt="Screenshot 2022-10-11 at 14 40 06" src="https://user-images.githubusercontent.com/52494172/195082917-93de54cd-4bf6-4355-8f37-f514b5b91218.png">




___
Also fixed this button being to close to the image:
<img width="435" alt="Screenshot 2022-10-11 at 14 46 53" src="https://user-images.githubusercontent.com/52494172/195083010-377a8c02-f585-4ae6-9800-632bda6cfabd.png">

___
and adjusted the space between some section making sure there are at least 40px between them **on Mobile:**
<img width="1270" alt="Screenshot 2022-10-11 at 14 48 07" src="https://user-images.githubusercontent.com/52494172/195083140-05b8ee97-e401-4804-829f-aa41ae3baf78.png">


We can deploy this later if it's hard to sync the changes from https://demosites.io/otter/patterns to the actual Otter build.

**Comment by JohnPixle** [comment]
@mghenciu absolutely, no need to think twice about making tweaks like that. Glad you like them.
Please go ahead and fix whatever you see can be improved. 

I think it will be a bit hard to sync the changes, this happens manually as far as I know. Maybe we can do it in the next revision? Or @HardeepAsrani can point us to the code in the build so we can insert the updated patterns code whenever we do a change? Let us know what you think it's the best way.

**Comment by mghenciu** [comment]
2 more things:
- @HardeepAsrani , I saw that we are using a **Row** block in the patterns. Is this one available in Wordpress 5.8, or what is our minor required wp version?

___
- With **Twenty Twenty-One**, looks like the Paragraph don't have bottom padding, @JohnPixle .
Do you think we should convert those to Adv. Headings? (I can do this) or add a spacer or just leave them like this:
<img width="1719" alt="Screenshot 2022-10-11 at 15 14 39" src="https://user-images.githubusercontent.com/52494172/195087356-c1d555f7-210a-4302-8021-c40e81dd1cc7.png">

**Comment by JohnPixle** [comment]
@mghenciu Regarding the paragraph spacing in TT1, I think we should leave things as they are to be honest. 

IF we convert some of our core paragraph blocks into Advanced Headings, we would need to do this for all our text instances, in order to be consistent across our patterns. Right now, only a few necessary Typographic elements are Advanced Headings.

If you want you can give it a try with a low-value spacer, I am open to that, but we might as well run into the issue where in other themes the distance will be too much. I think only TT1 does not add margins to paragraphs.

Let me know your thoughts!

**Comment by ineagu** [comment]
@JohnPixle as I mentioned before for me the problem with the image seems a challenge because as I see it if it looks bad in 2 out of 5 cases, this is 40% of the cases and once a user encounters an issue like this, I don't think he would trust to use the pattern, if we can get reliability and have more specific css as a tradeoff, this should make sense.

**Comment by JohnPixle** [comment]
@ineagu I appreciate the input here, I agree. I will sort it out with Hardeep on how to technically manage this. One solution would be to add custom CSS on the images, but this will have a cost in the customisability by the user, because in this case the core image alignment controls will not work. 

So if they want to import the centered-aligned testimonial and then left-align it manually, then it will not work (due to the custom CSS).

One other possible solution would be to left-align the elements patterns, but still in this case it depends on the theme. 
I will look into this and keep you posted.

@mghenciu regarding the paragraph margins: 
> I guess we can leave the spacing as it is for this release, and when doing a more in depth testing of Otter + other Top Themes (part of our quarterly goal) - we'll see how big of an issue this is or not

I agree, and in case we add a block gap control in Otter's sections columns in the future (like core groups have), then this will also affect the spacing.

**Comment by JohnPixle** [comment]
@HardeepAsrani Whenever you have some bandwidth, what are your thoughts on approaching the image alignment issue? The "problematic" themes in my [screenshot above](https://user-images.githubusercontent.com/4193672/194811541-12b33cd9-8f44-484c-b53a-4b55570f8384.png) are **TT2** and **Raft**. 

I am not concerned about Raft tbh because I will ask Andrei tomorrow to take a look into it and push a patch to fix it.

But for TT2, as seen in this page: https://patterns.s4-tastewp.com/patterns-test/ can you identify what the problem is?

In the editor, the image is set to centered alignment
![Screenshot 2022-10-11 at 18 57 31](https://user-images.githubusercontent.com/4193672/195141322-6412e18f-7bfb-445f-b5ea-ac67c2dc94cb.png)

If we were to force some specificity with custom CSS on the image block, what would that CSS be?

In case you need:

```
 Admin area URL: https://patterns.s4-tastewp.com/wp-admin 
 Username: pixle 
 Password: 5VU_pAajIVk 
```

Thanks!

**Comment by JohnPixle** [comment]
@HardeepAsrani Just a heads-up that Andrei has addressed the alignment issue in Raft in [this commit. ](https://github.com/Codeinwp/raft/commit/f91b701a7f7b5fc702ca17cbbd694f053cb7935c)

Should be up in the next Raft update.

**Comment by ineagu** [comment]
@HardeepAsrani it seems to be from the fact that some themes are not adding the styles for those classes for images. Is not this something that we can add in Otter as a fallback? to style those things?

**Comment by HardeepAsrani** [comment]
@ineagu Yea, I've added that and it should work fine with those themes now.

**Comment by HardeepAsrani** [comment]
@mghenciu @JohnPixle I tried and it seems there are few issues with patterns in 5.8 and 5.9.

Few specific:

- For some reason, Icon Blocks in Patterns break in 5.9 or before, can't see what's causing it. The code is literally identical.
- The Spacer doesn't work either in 5.9 or before as the only in 6.0 it started to support vh unit.

Here are a few things we can do:

- I can convert these patterns as if they were made in WordPress 5.9 in which case we might not be able to use some new features, but it will work with the latest version of WordPress. Old-to-new transition always works.
- To add more, we can have a supported attribute in Patterns so we know the minimum version a particular pattern works in which way we can only register a particular pattern on the version that it works nicely on.


What do you think?

**Comment by HardeepAsrani** [comment]
We can discuss this in the meeting on Monday. I've identified a few issues. Nothing is too breaking. Pain points are:

- Core's Spacing Block
- ThemeIsle Icons in Icon Block or Icons List

Once we agree, I can push a PR within hours to fix it, and we can have a release on Tuesday/Wednesday.

**Comment by JohnPixle** [comment]
@HardeepAsrani 
I just tested the patterns with the latest changes, all is good. The CSS fix for the images works great now, thanks for taking care of it.

I only spotted an alignment control that was not properly set, on the social icons block at author box pattern (screenshot below)

![Screenshot 2022-10-18 at 14 17 20](https://user-images.githubusercontent.com/4193672/196415297-3d1026a7-a440-4620-bef3-8f3856ec0109.png)

I fixed it and I am attaching the json (I zipped it because GH does not allow json uploads).

[author.zip](https://github.com/Codeinwp/otter-blocks/files/9809684/author.zip)


Let me know if there is anything I can do at my end for this release.
